src/ngraph/pass/constant_folding_non_zero.cpp

```diff
@@ -0,0 +1,198 @@
+
+template <typename T>
+static shared_ptr<op::Constant> fold_constant_non_zero(const shared_ptr<op::Constant>& data)
+{
+    const auto input_shape = data->get_shape();
+    const auto* input_values = data->get_data_ptr<T>();
+    if (ngraph::is_scalar(input_shape))
+    {
+        const auto scalar_value = input_values[0];
+
+        NGRAPH_CHECK(scalar_value != T{0},
+                     "It's not possible to constant fold a NonZero op for a scalar equal to zero.");
+
+        // return 0(the only index) if the data input contains a scalar different than zero
+        return op::Constant::create(element::i64, Shape{1, 1}, {0});
+    }
+    else if (is_vector(input_shape))
+    {
+        const auto input_values_count = shape_size(input_shape);
+        std::vector<int64_t> indices;
+        indices.reserve(input_values_count);
+
+        const T zero_value = T{0};
+        for (size_t i = 0; i < input_values_count; ++i)
+        {
+            if (input_values[i] != zero_value)
+            {
+                indices.push_back(i);
+            }
+        }
+
+        indices.shrink_to_fit();
+        return op::Constant::create(element::i64, Shape{1, indices.size()}, indices);
+    }
+    else
+    {
+        NonZeroElements non_zero_elems{input_shape};
+        const auto& found_indices = non_zero_elems.find_indices(input_values);
+
+        // we can't create an empty Constant indicating that no non-zero elems were found
+        NGRAPH_CHECK(
+            found_indices.front().size() > 0,
+            "It's not possible to constant fold a NonZero op with an input containing only zeros.");
+
+        // flatten the results and return them as a Constant
+        std::vector<int64_t> indices;
+        indices.reserve(found_indices.size() * found_indices.front().size());
+        for (const auto& row : found_indices)
+        {
+            indices.insert(indices.end(), row.begin(), row.end());
+        }
+
+        const Shape out_shape{found_indices.size(), found_indices.front().size()};
+        return op::Constant::create(element::i64, out_shape, indices);
+    }
+}
+
+void pass::ConstantFolding::construct_constant_non_zero()
+{
+    const auto data_label = make_shared<pattern::op::Label>(
+        element::f32, Shape{2, 2, 3}, pattern::has_class<op::Constant>());
+    const auto non_zero = make_shared<op::v3::NonZero>(data_label);
+
+    auto constant_non_zero_callback = [data_label](pattern::Matcher& m) {
+        auto pattern_map = m.get_pattern_map();
+
+        const auto data = static_pointer_cast<op::Constant>(pattern_map[data_label]);
+
+        // const auto found_nz_node = m.get_match_root();
+        // this fails because the output of NonZero is still dynamic - is it ok to remove this line?
+        // or should the output shape of this op be calculated as the maximum possible number
+        // of non-zero indices it can return if the input is a constant?
+        // NGRAPH_CHECK(revalidate_and_ensure_static(found_nz_node));
```

Comment:
@diyessi I think this line should be removed for an op that is dynamic until the end of its inference. Is this correct? Please have a look at the comment I left in this code above.